File: leetcode/203-remove-linked-list-elements/main.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

// classic solution
// time		O(n)
// space 	O(1)
// beats 20% (i think leetcode has a problem in computing the time)
func removeElements(head *ListNode, val int) *ListNode {
	dump := &ListNode{0, head}
	cur := dump
	for cur.Next != nil {
		if cur.Next.Val == val {
			cur.Next = cur.Next.Next
		} else {
			cur = cur.Next
		}
	}
	return dump.Next
}

func main() {

}

File: leetcode/203-remove-linked-list-elements/main.py
"""
    1st approach
    - remove the next pointer if it contains the target in a loop

    Time    (n)
    Space   O(1)
    56 ms, faster than 100.00%
"""


class Solution(object):
    def removeElements(self, head, val):
        """
        :type head: ListNode
        :type val: int
        :rtype: ListNode
        """
        dump = ListNode(-1)
        dump.next = head
        cur = dump
        while cur != None:
            while cur.next != None and cur.next.val == val:
                cur.next = cur.next.next
            cur = cur.next
        return dump.next


"""
    classic approach
    - remove the next pointer if it contains the target

    Time    (n)
    Space   O(1)
    56 ms, faster than 100.00%
"""


class Solution(object):
    def removeElements(self, head, val):
        """
        :type head: ListNode
        :type val: int
        :rtype: ListNode
        """
        dump = ListNode(0)
        dump.next = head
        prev = dump
        cur = head
        while cur != None:
            if cur.val == val:
                prev.next = cur.next
            else:
                prev = cur
            cur = cur.next
        return dump.next
